File: summer_2024/OOP - Diffie-Hellman/Python/Field.py
    
P = 1234577

class Field:

    def _numberToField(self, x):
        return ((x % P) + P) % P
        
    def _inverse(self, a):
        m = P
        x = 1
        y = 0

        while a > 1:
            quotient = a // m
            t = m

            m = a % m
            a = t
            t = y

            y = x - quotient * y
            x = t

        if x < 0:
            x += P

        return x

    def __init__(self,value=0):
        self.value = self._numberToField(value)

    def __assign__(self, obj):
        self = obj

    def __add__(self, obj):
        res = Field()
        res.value = res._numberToField(self._numberToField(self.value) + obj._numberToField(obj.value))
        return res

    def __sub__(self, obj):
        res = Field()
        res.value = res._numberToField(self._numberToField(self.value) - obj._numberToField(obj.value))
        return res

    def __mul__(self, obj):
        res = Field()
        res.value = res._numberToField(self._numberToField(self.value) * obj._numberToField(obj.value))
        return res


    def __truediv__(self, obj):
        if obj.value == 0:
            raise ZeroDivisionError("Nie można obliczyć odwrotności 0 w ciele")
        res = Field()
        res.value = res._numberToField(self._numberToField(self.value) * obj._inverse(obj.value))
        return res

    def __floordiv__(self, obj):
        res = Field()
        res.value = res._numberToField(self._numberToField(self.value) * obj._inverse(obj.value))
        return res

    def __iadd__(self, obj):
        self.value = self._numberToField(self._numberToField(self.value) + obj._numberToField(obj.value))
        return self

    def __isub__(self, obj):
        self.value = self._numberToField(self._numberToField(self.value) - obj._numberToField(obj.value))
        return self

    def __imul__(self, obj):
        self.value = self._numberToField(self._numberToField(self.value) * obj._numberToField(obj.value))
        return self

    def __idiv__(self, obj):
        self.value = self._numberToField(self._numberToField(self.value) * obj._inverse(obj.value))
        return self

    def __ifloordiv__(self, obj):
        self.value = self._numberToField(self._numberToField(self.value) * obj._inverse(obj.value))
        return self
    
    def __eq__(self,obj):
        if(self.value == obj.value):
            return True
        return False
    
    def __ne__(self,obj):
        if(self.value != obj.value):
            return True
        return False

    
    def __ge__(self,obj):
        if(self.value >= obj.value):
            return True
        return False

    
    def __le__(self,obj):
        if(self.value <= obj.value):
            return True
        return False

    
    def __lt__(self,obj):
        if(self.value < obj.value):
            return True
        return False
    
    def __gt__(self,obj):
        if(self.value > obj.value):
            return True
        return False
    
    def getCharacteristics(self):
        return P
    
    def getValue(self):
        return self.value
    

File: summer_2024/OOP - Diffie-Hellman/Java/Field.java
public class Field {
    
    private final static long P = 1234577;
    private long value;
    
    private long numberToField(long x){
        return ((x % P) + P) % P;
    }

    private long inverse(long a){
        long m = P;
        long x = 1;
        long y = 0;

        while(a > 1) {
            long quotient = a / m;
            long t = m;

            m = a % m;
            a = t;
            t = y;

            y = x - quotient * y;
            x = t;
        }

        if(x < 0)
            x += P;

        return x;
    }

    private Field(){
        value = 0;
    }
    
    public Field(long val){
        value = numberToField(val);
    }

    public long getValue(){
        return value;
    }

    public static long getCharacteristic(){
        return P;
    }

    public Field add(Field b){
        Field result = new Field();
        result.value = numberToField(numberToField(value) + numberToField(b.value));
        return result; 
    }
    
    public Field sub(Field b){
        Field result = new Field();
        result.value = numberToField(numberToField(value) - numberToField(b.value));
        return result; 
    }

    public Field mul(Field b){
        Field result = new Field();
        result.value = numberToField(numberToField(value) * numberToField(b.value));
        return result; 
    }

    public Field div(Field b) throws ArithmeticException{
        if(numberToField(b.value) == 0){
            throw new ArithmeticException("Nie można obliczyć odwrotności 0 w ciele");
        }
        Field result = new Field();
        result.value = numberToField(numberToField(value) * inverse(b.value));
        return result; 
    }

    public void assign(Field b){
        this.value = numberToField(b.value);
    }

    public static boolean eq(Field a, Field b){
        return a.value == b.value;
    }

    public static boolean neq(Field a, Field b){
        return a.value != b.value;
    }

    public static boolean geq(Field a, Field b){
        return a.value >= b.value;
    }

    public static boolean leq(Field a, Field b){
        return a.value <= b.value;
    }

    public static boolean gt(Field a, Field b){
        return a.value > b.value;
    }

    public static boolean lt(Field a, Field b){
        return a.value < b.value;
    }
}
